experimental/lib/Transforms/Speculation/PlacementFinder.cpp

```diff
@@ -0,0 +1,359 @@
+      if (specValues.count(operand))
+        hasSpecInput = true;
+      else
+        hasNonSpecInput = true;
+    }
+
+    if (hasSpecInput and hasNonSpecInput) {
+      for (Value operand : blockOp->getOperands()) {
+        // Create a Save for every non-speculative operand
+        if (not specValues.count(operand)) {
+          // No save needed in front of Source Operations
+          if (isa<handshake::SourceOp>(operand.getDefiningOp()))
+            continue;
+
+          placements.addSave(operand, blockOp);
+        }
+      }
+    }
+  }
+
+  return success();
+}
+
+//===----------------------------------------------------------------------===//
+// Commit Units Finder Methods
+//===----------------------------------------------------------------------===//
+
+void PlacementFinder::findCommitsTraversal(llvm::DenseSet<Operation *> &visited,
+                                           Operation *currOp) {
+  // End traversal if currOp is already in visited set
+  if (auto [_, isNewOp] = visited.insert(currOp); !isNewOp)
+    return;
+  for (Value res : currOp->getResults()) {
+    for (Operation *succOp : res.getUsers()) {
+      if (placements.containsSave(res, succOp)) {
+        // A Commit is needed in front of Save Operations. This will be later
+        // converted into a SaveCommit for multiple loop speculation.
+        placements.addCommit(res, succOp);
+      } else if (isa<handshake::LSQOp, handshake::MemoryControllerOp,
+                     handshake::MCStoreOp, handshake::LSQStoreOp,
+                     handshake::MCLoadOp, handshake::LSQLoadOp>(succOp)) {
+        // A commit is needed in front of memory operations
+        placements.addCommit(res, succOp);
+      } else if (isa<handshake::EndOp>(succOp)) {
+        // A commit is needed in front of the end/exit operation
+        placements.addCommit(res, succOp);
+      } else {
+        findCommitsTraversal(visited, succOp);
+      }
+    }
+  }
+}
+
+struct endpointComparator {
+  bool operator()(const BBEndpoints &a, const BBEndpoints &b) const {
+    if (a.srcBB != b.srcBB)
+      return a.srcBB < b.srcBB;
+    return a.dstBB < b.dstBB;
+  }
+};
+using BBEndpointsMap =
+    std::map<BBEndpoints, std::vector<OpPlacement>, endpointComparator>;
+
+// Data structure to hold all arcs leading to a single BB predecessor
+// Note: srcBB and dstBB can be equal when the arcs are Backedges
+struct BlockPredecessor {
+  unsigned srcBB;
+  unsigned dstBB;
+  std::vector<OpPlacement> arcs;
+};
+using BBPredecessorMap = std::map<unsigned, std::vector<BlockPredecessor>>;
+
+// Get a map from BBs to the list of its BlockPredecessors
+static BBPredecessorMap getBlockPredecessors(handshake::FuncOp funcOp) {
+  // Traverse all operations to find arcs between BBs (and self-arcs)
+  BBEndpointsMap endpointArcs;
+  funcOp->walk([&](Operation *op) {
+    for (Value operand : op->getOperands()) {
+      BBEndpoints endpoints;
+      if (isBackedge(operand, op, &endpoints) or
```

Comment:
why are the endpoints being set by isBackedge? Is this normal? it looks very strange

Reply:
The function isBackedge is overloaded. If a reference to a BBEndpoints instance is passed, the endpoints of the edge (operand, op) is set. This allows for the later comparison in the same line `endpoints.srcBB != endpoints.dstBB`. This is according to the design of the `CFG.h` API.